### ISSUE TITLE
Copy .crt files to temp-file if executed from phar

### ISSUE THREAD

**Issue by sebastianfeldmann** [dropbox/dropbox-sdk-php#26]
If the lib is compiled into a phar the setting **CURLOP_CAINFO** and **CURLOPT_CAPATH** will fail.

In this case the lib should move the .crt file to a temp file and use this for current curl request.

**Comment by cakoose** [comment]
I'm trying to figure out how to:
1. Create the temp file reliably and securely.
2. Only create the temp file the first time.

If I use `tmpfile()`, I think it is reliable and secure (though maybe `mkstemp()` would be better) but the name of the file isn't predictable so I can't ensure that the file hasn't already been created.

If I use `sys_get_temp_dir()`, I can easily check if the file has already been created, but I don't think this method is as robust (what if the file has been created by some other instance) or secure.

Any advice?

**Comment by sebastianfeldmann** [comment]
I just created a pull request for this Issue.
I used `sys_get_temp_dir()` to create the file.

**Comment by cakoose** [comment]
I have some concerns about edge cases when using a single global file path.

For example, what if two apps are using different versions of the Dropbox PHP SDK, with different ".crt" files.  Will they overwrite each other's ".crt" files?

Is `copy()` atomic?  What if two requests come in at the same time and both request-handling threads try to perform the `copy()` at around the same time?

**Comment by antoncohen** [comment]
Is it possible to use a file-like memory stream instead of a path to a file on disk? Something like: https://bugs.php.net/bug.php?id=43468

**Comment by sebastianfeldmann** [comment]
Good points
`copy` ist not atomic, the only atomic alternative i know is `rename`.
So maybe copy to a unique filename and rename to the configured filename?

Different versions with different .crt files could be fixed with a version numbered filenames.
But there is the hassle with the versioning.

Creating unique files for each request would be a huge overhead if the PHAR really is getting under fire.
I don't know if the PHAR scenario is likely to be used in a high load scenario, i more likely doubt it.

**Comment by sebastianfeldmann** [comment]
@antoncohen when i remember correctly you could get something out into a stream but you can't get anything in this way.
When i filed this curl PHAR bug at php.net recently the answer was something like.
CURL is an external lib and it knows nothing about PHARs oder streams, you can't even pass in a string.

**Comment by cakoose** [comment]
What are some of the reasons people use PHARs for website deployments?  Is the whole app in a single PHAR or is the SDK in its own PHAR?

Is there's something about the scenario that we can take advantage of?  For example, maybe we can have special instructions for PHAR-based deployments (e.g: extract the .crt file and place it next to the .phar file).

The PHP streams API supports SSL (http://php.net/manual/en/context.ssl.php).  I wonder if the `cainfo` option can take a stream.  The docs make me feel like they don't, but maybe it's worth checking.  (Switching from CURL to streams would be a larger task, though.)

**Comment by sebastianfeldmann** [comment]
I tried to use `stream`, sadly it didn't work.

For me a PHAR is a "don't worry be happy solution" so i don't want to hassle with any deployment or pre setup to make something work.

You could offer an API so everybody could handle the extraction by themselves if they want to.
Something like:
```php
Interface PharCacertExtractor
{
    public function extract($pathToCacert);
```
Provide a default `Extractor` with some similar code as mine, but offer the Possibility to set a custom `Extractor` like
```php
$client = new dbx\Client($token, "somekey");
// Client::setCacertExtractor(PharCacertExtractor $extractor);
$client->setCacertExtractor($myExtractor);
```

So if someone is getting into heavy concurrency problems he can handle them on his own.

**Comment by sebastianfeldmann** [comment]
In my case i ship a little app that bundles all required libs within a single PHAR.
Just like phpunit.phar, phpcs.phar, phpab.phar... or many other of those PHAR-apps.
Personally i would not deploy PHARs for serving a web application, but thats only me ;)

**Comment by cakoose** [comment]
Ah ok, so this is a command-line tool?  This specific use case seems easier to deal with.  If it's ok to extract the ".crt" file once per run of the tool, I think we can just use `tmpfile()`.

**Comment by sebastianfeldmann** [comment]
Would be absolutly fine for me ;)

**Comment by sebastianfeldmann** [comment]
Should i update my code to use `tmpfile()` and send another pull request or do you want to code it?

**Comment by cakoose** [comment]
I'll take a stab at it this week.

Thanks a lot for helping me understand the issue!

**Comment by cakoose** [comment]
I've pushed to a branch: https://github.com/dropbox/dropbox-sdk-php/tree/phar

Can you pull from there and see if it works for you?  If it does, I'll push it to master.

**Comment by sebastianfeldmann** [comment]
Pull incoming

**Comment by sebastianfeldmann** [comment]
I added
```php
if (substr(__FILE__, 0, 7) == 'phar://') {
    Dropbox\RootCertificates::useExternalPaths();
}
```
to my code and it worked like a charm

**Comment by cakoose** [comment]
Fixed in 673985f549eba8f52e258a49e8c24840a6fb10ea

**Comment by sebastianfeldmann** [comment]
Any plans for a new version so I could get this fix via composer?
Thanks for the fast fix.
Cheers, Sebastian

**Comment by cakoose** [comment]
Just released v1.1.5.
